### PR TITLE
[PSFP1-0004][PFSP1-0008] Profile Landing Screen, Profile Detail

### LINKED ISSUE #23: [PFSP1-0004] Profile Landing Screen
### Priority 
- [ ] Low
- [x] Mid
- [ ] High

### Links
Name | Link
-|-
User Story | [TBD]()
Figma | [link](https://www.figma.com/file/YWJimHMl9fq3v4dQBZSuQo/Hi-Fi-User?node-id=388%3A2)

### Completion
Expected | Time
-|-
Expected Start | 27-07-2020
Expected End | 29-07-2020

### LINKED ISSUE #29: [PFSP1-0008] Profile Detail
### Priority 
- [x] Low
- [ ] Mid
- [ ] High

### Links
Name | Link
-|-
User Story | [link]()
Figma | [link](https://www.figma.com/file/YWJimHMl9fq3v4dQBZSuQo/Hi-Fi-User?node-id=393%3A4930)

### Completion
Expected | Time
-|-
Expected Start | 29-07-2020
Expected End | 03-07-2020

### PR DESCRIPTION
**Pull request by SetyaG**
Closes #23 
Closes #29
***
## Description
Create the profile landing page and edit profile page, navigate from profile landing page to edit profile page

## Related Links
List related PRs against other branches:

List | link
------ | ------
User Story | [TBD]()
Figma | [link](https://www.figma.com/file/YWJimHMl9fq3v4dQBZSuQo/Hi-Fi-User?node-id=388%3A2)
Issue | [link](https://github.com/inxudianz/Project-Famili/issues/23)

## Todos
- [X] UI
- [X] Coordinator
- [ ] Network
- [ ] Tests
- [ ] Documentation

## Screenshot / Video
Screenshot consisting of changes made on this branch.

### UI Changes
UI related changes will be shown here (can be either screenshot or video format)

### Network Status
Network related changes will be shown here (screenshot of request status)

### Test Coverage
Test coverage result for designated changes will be shown here (screenshot of coverage)

## Notes
The **text fields** and **button** has not changed to custom text field and button

### CODE REVIEW

File: MainApp/MainApp/Coordinator/Feature/Profile/EditProfileCoordinator.swift [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+class EditProfileCoordinator: EditProfileCoordinatorProtocol {
+    
+    func saveEditProfile() {
+        let vc = ProfileLandingController()
+        let vm = ProfileLandingViewModel()
+        vm.view = vc
+        vc.viewModel = vm
+        navigationController?.pushViewController(vc, animated: false)
+    }
```
Comment (inxudianz) [inline MainApp/MainApp/Coordinator/Feature/Profile/EditProfileCoordinator.swift]: variable first then function please

File: MainApp/MainApp/Coordinator/Feature/Profile/EditProfileCoordinator.swift [hunk 2]
```diff
@@ -0,0 +1,43 @@
+//
+//  EditProfileCoordinator.swift
+//  MainApp
+//
+//  Created by Owen Prasetya on 03/08/20.
+//  Copyright © 2020 William Inx. All rights reserved.
+//
+
+import Foundation
```
Comment (inxudianz) [inline MainApp/MainApp/Coordinator/Feature/Profile/EditProfileCoordinator.swift]: no need to import foundation

File: MainApp/MainApp/Coordinator/Feature/Profile/ProfileLandingCoordinator.swift [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import UIKit
+
+class ProfileLandingCoordinator: ProfileLandingCoordinatorProtocol {
+    func getEditProfile() {
+        let vc = EditProfileController()
+        let vm = EditProfileViewModel()
+        vm.view = vc
+        vc.viewModel = vm
+        navigationController?.pushViewController(vc, animated: false)
+    }
```
Comment (inxudianz) [inline MainApp/MainApp/Coordinator/Feature/Profile/ProfileLandingCoordinator.swift]: variable first then functions

File: MainApp/MainApp/Coordinator/Feature/Profile/ProfileLandingCoordinator.swift [hunk 4]
```diff
@@ -0,0 +1,47 @@
+//
+//  ProfileLandingCoordinator.swift
+//  MainApp
+//
+//  Created by Owen Prasetya on 29/07/20.
+//  Copyright © 2020 William Inx. All rights reserved.
+//
+
+import Foundation
```
Comment (inxudianz) [inline MainApp/MainApp/Coordinator/Feature/Profile/ProfileLandingCoordinator.swift]: remove extra import

File: MainApp/MainApp/Coordinator/Feature/Profile/ProfileLandingCoordinator.swift [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        navigationController.navigationItem.title = "Welcome, User!"
+    }
+    
+    func start() {
+        let vc = ProfileLandingController()
+        let vm = ProfileLandingViewModel()
+        vc.navigationItem.configure()
+        vc.navigationController?.navigationBar.prefersLargeTitles = true
+        vc.navigationItem.largeTitleDisplayMode = .always
+        vc.navigationItem.title = "Welcome, User!"
```
Comment (inxudianz) [inline MainApp/MainApp/Coordinator/Feature/Profile/ProfileLandingCoordinator.swift]: Kindly create constant swift file containing String literal or any kind of literal. example can be viewed in sample or component

File: MainApp/MainApp/Feature/Profile/CustomTableViewCell.swift [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+//        view.frame = bounds
+//        addSubview(view)
+//    }
+//
+//    func loadViewFromNib() -> UIView {
+//        let bundle = Bundle(for: type(of: self))
+//        let nib = UINib(nibName: "CustomTableViewCell", bundle: bundle)
+//        let view = nib.instantiate(withOwner: self, options: nil)[0] as! UIView
+//        return view
+//    }
```
Comment (inxudianz) [inline MainApp/MainApp/Feature/Profile/CustomTableViewCell.swift]: remove dead code

File: MainApp/MainApp/Feature/Profile/CustomTableViewCell.swift [hunk 7]
```diff
@@ -0,0 +1,40 @@
+//
+//  CustomTableViewCell.swift
+//  MainApp
+//
+//  Created by Owen Prasetya on 30/07/20.
+//  Copyright © 2020 William Inx. All rights reserved.
+//
+
+import Foundation
```
Comment (inxudianz) [inline MainApp/MainApp/Feature/Profile/CustomTableViewCell.swift]: remove foundation

File: MainApp/MainApp/Feature/Profile/EditProfileController.swift [hunk 8]
```diff
@@ -0,0 +1,27 @@
+//
+//  EditProfileController.swift
+//  MainApp
+//
+//  Created by Owen Prasetya on 03/08/20.
+//  Copyright © 2020 William Inx. All rights reserved.
+//
+
+import Foundation
```
Comment (inxudianz) [inline MainApp/MainApp/Feature/Profile/EditProfileController.swift]: remove foundation

File: MainApp/MainApp/Feature/Profile/EditProfileController.swift [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+import UIKit
+import Component
+
+class EditProfileController: MasterViewController, EditProfileProtocol {
+    var viewModel: EditProfileViewModelProtocol?
```
Comment (inxudianz) [inline MainApp/MainApp/Feature/Profile/EditProfileController.swift]: outlet first then variable

File: MainApp/MainApp/Feature/Profile/EditProfileViewModel.swift [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+    
+    weak var view: EditProfileProtocol?
+    
+    weak var coordinator: EditProfileCoordinatorProtocol?
+    
+    func navigateToProfile() {
+        coordinator?.saveEditProfile()
+    }
+    
+
```
Comment (inxudianz) [inline MainApp/MainApp/Feature/Profile/EditProfileViewModel.swift]: remove extra white spaces

File: MainApp/MainApp/Feature/Profile/ProfileLandingController.swift [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+//  Created by Owen Prasetya on 29/07/20.
+//  Copyright © 2020 William Inx. All rights reserved.
+//
+
+import UIKit
+import Component
+
+class ProfileLandingController: MasterViewController, ProfileLandingViewProtocol, UITableViewDelegate, UITableViewDataSource {
+    
+    var viewModel: ProfileLandingViewModelProtocol?
```
Comment (inxudianz) [inline MainApp/MainApp/Feature/Profile/ProfileLandingController.swift]: outlet, variable, function

File: MainApp/MainApp/Feature/Profile/ProfileLandingController.swift [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+//  MainApp
+//
+//  Created by Owen Prasetya on 29/07/20.
+//  Copyright © 2020 William Inx. All rights reserved.
+//
+
+import UIKit
+import Component
+
+class ProfileLandingController: MasterViewController, ProfileLandingViewProtocol, UITableViewDelegate, UITableViewDataSource {
```
Comment (inxudianz) [inline MainApp/MainApp/Feature/Profile/ProfileLandingController.swift]: refactor tableviewDelegate and datasource to its own file, and get the data from viewModel

File: MainApp/MainApp/Feature/Profile/ProfileLandingController.swift [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+    }
+    
+    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
+        switch section {
+        case 0:
+            return "Account"
+        case 1:
+            return "General"
+        default:
+            return ""
```
Comment (inxudianz) [inline MainApp/MainApp/Feature/Profile/ProfileLandingController.swift]: create constant file like above

File: MainApp/MainApp/Feature/Profile/ProfileLandingViewModel.swift [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+import Foundation
+
+class ProfileLandingViewModel: ProfileLandingViewModelProtocol {
+    func navigateToHelp() {
+        // Call coordinator to Help
+    }
+    
+    func navigateToFreqAskedQuestion() {
+        // Call coordinator to FAQ
+    }
```
Comment (inxudianz) [inline MainApp/MainApp/Feature/Profile/ProfileLandingViewModel.swift]: variable - > function

File: MainApp/MainApp/Coordinator/Feature/Profile/EditProfileCoordinator.swift [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+import UIKit
+
+class EditProfileCoordinator: EditProfileCoordinatorProtocol {
```
Comment (inxudianz) [inline MainApp/MainApp/Coordinator/Feature/Profile/EditProfileCoordinator.swift]: please combine editProfileCoordinator and profileLandingCoordinator as it is one same feature. name it ProfileCoordinator

File: MainApp/MainApp.xcodeproj/project.pbxproj [hunk 16]
```diff
@@ -313,14 +343,28 @@
...
 		};
 		923249A324C03FF600FDFC3E /* Profile */ = {
 			isa = PBXGroup;
 			children = (
+				1EBF012324D0A6130098C24A /* ProfileLandingController.xib */,
+				1EBF012924D0A65F0098C24A /* ProfileLandingController.swift */,
+				1EBF012B24D0A6A00098C24A /* ProfileLandingViewModel.swift */,
+				1EBF012D24D0A6B00098C24A /* ProfileLandingProtocol.swift */,
+				1EBF014924D1E5450098C24A /* CustomTableViewCell.swift */,
+				1EBF014B24D1E5510098C24A /* CustomTableViewCell.xib */,
```
Comment (inxudianz) [inline MainApp/MainApp.xcodeproj/project.pbxproj]: Please seperate custom cell into its own group

File: MainApp/MainApp.xcodeproj/project.pbxproj [hunk 17]
```diff
@@ -313,14 +343,28 @@
...
 				923249AB24C0406B00FDFC3E /* Resources */,
+				1E44DC5124D855C600025AE8 /* EditProfileController.swift */,
+				1E44DC5524D8560400025AE8 /* EditProfileController.xib */,
+				1E44DC5724D8562D00025AE8 /* EditProfileProtocol.swift */,
```
Comment (inxudianz) [inline MainApp/MainApp.xcodeproj/project.pbxproj]: please create group inside feature/profile called edit and put the edit related stuff inside

**Review by inxudianz — CHANGES_REQUESTED** [review]

File: MainApp/MainApp/FeatureNetwork/ProfileNetwork/ProfileNetwork.swift [hunk 18]
```diff
@@ -8,6 +8,68 @@
...
+    func didFailedEditProfile(error: Error)
+}
+
+protocol ProfileLandingNetworkProtocol {
+    var profileLandingNetworkDelegate: ProfileLandingNetworkDelegate? { get set }
+    
+    func profileGet()
+}
+
+protocol EditProfileNetworkProtocol {
```
Comment (inxudianz) [inline MainApp/MainApp/FeatureNetwork/ProfileNetwork/ProfileNetwork.swift]: combine profileLanding and editProfile to ProfileNetwork

**Review by inxudianz — CHANGES_REQUESTED** [review]

File: MainApp/MainApp/Feature/Profile/ProfileLanding/ProfileLandingController.swift [hunk 19]
```diff
@@ -11,20 +11,29 @@ import Component
 
 class ProfileLandingController: MasterViewController, ProfileLandingViewProtocol, UITableViewDelegate, UITableViewDataSource {
```
Comment (inxudianz) [inline MainApp/MainApp/Feature/Profile/ProfileLanding/ProfileLandingController.swift]: datasource should be handled outside of view controller

**Review by inxudianz — CHANGES_REQUESTED** [review]

File: MainApp/MainApp/Coordinator/Feature/Profile/EditProfileCoordinator.swift (hunk 1 above)
Comment (SetyaG) [reply MainApp/MainApp/Coordinator/Feature/Profile/EditProfileCoordinator.swift]: done

File: MainApp/MainApp/FeatureNetwork/ProfileNetwork/ProfileNetwork.swift (hunk 18 above)
Comment (SetyaG) [reply MainApp/MainApp/FeatureNetwork/ProfileNetwork/ProfileNetwork.swift]: combined

File: MainApp/MainApp/Feature/Profile/ProfileLanding/ProfileLandingController.swift (hunk 19 above)
Comment (SetyaG) [reply MainApp/MainApp/Feature/Profile/ProfileLanding/ProfileLandingController.swift]: datasource moved to view model

File: MainApp/MainApp.xcodeproj/project.pbxproj (hunk 17 above)
Comment (SetyaG) [reply MainApp/MainApp.xcodeproj/project.pbxproj]: Moved to its own folder

File: MainApp/MainApp.xcodeproj/project.pbxproj (hunk 16 above)
Comment (SetyaG) [reply MainApp/MainApp.xcodeproj/project.pbxproj]: Moved to cell to its own folder

File: MainApp/MainApp/Coordinator/Feature/Profile/EditProfileCoordinator.swift (hunk 15 above)
Comment (SetyaG) [reply MainApp/MainApp/Coordinator/Feature/Profile/EditProfileCoordinator.swift]: Combined the coordinator and protocol

File: MainApp/MainApp/Feature/Profile/ProfileLandingController.swift (hunk 13 above)
Comment (SetyaG) [reply MainApp/MainApp/Feature/Profile/ProfileLandingController.swift]: Created constant file

File: MainApp/MainApp/Feature/Profile/ProfileLandingViewModel.swift (hunk 14 above)
Comment (SetyaG) [reply MainApp/MainApp/Feature/Profile/ProfileLandingViewModel.swift]: Done

File: MainApp/MainApp/Coordinator/Feature/Profile/ProfileLandingCoordinator.swift (hunk 5 above)
Comment (SetyaG) [reply MainApp/MainApp/Coordinator/Feature/Profile/ProfileLandingCoordinator.swift]: Created constant

File: MainApp/MainApp/Coordinator/Feature/Profile/ProfileLandingCoordinator.swift (hunk 3 above)
Comment (SetyaG) [reply MainApp/MainApp/Coordinator/Feature/Profile/ProfileLandingCoordinator.swift]: done

File: MainApp/MainApp/Coordinator/Feature/Profile/ProfileLandingCoordinator.swift (hunk 4 above)
Comment (SetyaG) [reply MainApp/MainApp/Coordinator/Feature/Profile/ProfileLandingCoordinator.swift]: done

**Review by inxudianz — APPROVED** [review]

**Pull request merged**